### PR TITLE
scheduler: compatibility about query info in hot scheduler

### PR DESCRIPTION
**Pull request by lhy1024**
### What problem does this PR solve?

Low versions of tikv do not have qps statistics, so when using a low version of tikv with a high version of pd, the hotspot scheduling will not be as expected

### What is changed and how it works?

For upgrading from versions before 5.2 to 5.2, regardless of the read priorities and write priorities, pd will schedule hotspots according to byte and key priorities until all tikv's in the cluster have been upgraded.

### Check List

Tests

- Unit test

### Release note

```release-note
None
```

### CODE REVIEW

**Review by Yisaer — COMMENTED** [review]
During upgrading tikv, pd-ctl may still show `qps` as default strategy by `scheduler config`

**Comment by HunDunDM** [comment]
Is it enough to explain in the documentation?

**Review by Yisaer — APPROVED** [review]

File: server/schedulers/hot_region.go [hunk 1]
```diff
@@ -422,11 +423,9 @@ type balanceSolver struct {
 	minDst   *storeLoad
 	rankStep *storeLoad
 
-	firstPriority             int
-	secondPriority            int
-	writeLeaderFirstPriority  int
-	writeLeaderSecondPriority int
-	isSelectedDim             func(int) bool
+	firstPriority  int
+	secondPriority int
```
Comment (HunDunDM) [inline server/schedulers/hot_region.go:427]: Add comments to explain that it is related to `statistics.DimLen`.

File: server/schedulers/hot_region.go [hunk 2]
```diff
@@ -489,17 +488,34 @@ func (bs *balanceSolver) init() {
...
+		bs.firstPriority, bs.secondPriority = bs.adjustConfig(bs.sche.conf.GetReadPriorities(), []string{BytePriority, KeyPriority})
 	} else {
-		bs.firstPriority, bs.secondPriority = prioritiesTodim(bs.sche.conf.WritePeerPriorities)
-		bs.writeLeaderFirstPriority, bs.writeLeaderSecondPriority = prioritiesTodim(bs.sche.conf.WriteLeaderPriorities)
+		if bs.opTy == transferLeader {
+			bs.firstPriority, bs.secondPriority = bs.adjustConfig(bs.sche.conf.GetWriteLeaderPriorites(), []string{KeyPriority, BytePriority})
+		} else {
+			bs.firstPriority, bs.secondPriority = bs.adjustConfig(bs.sche.conf.GetWritePeerPriorites(), []string{BytePriority, KeyPriority})
+		}
 	}
```
Comment (HunDunDM) [inline server/schedulers/hot_region.go:498]: Is it better to use `switch case`?

File: server/schedulers/hot_region.go [hunk 3]
```diff
@@ -608,7 +608,7 @@ func (bs *balanceSolver) filterSrcStores() map[uint64]*storeLoadDetail {
 }
 
 func (bs *balanceSolver) checkSrcByDimPriorityAndTolerance(minLoad, expectLoad *storeLoad) bool {
-	if bs.sche.conf.StrictPickingStore {
```
Comment (HunDunDM) [inline server/schedulers/hot_region.go:595]: use `GetStrictPickingStore()`

**Review by HunDunDM — COMMENTED** [review]
rest LGTM

**Review by HunDunDM — APPROVED** [review]

File: server/schedulers/hot_region.go [hunk 4]
```diff
@@ -435,11 +436,11 @@ type balanceSolver struct {
 	minDst   *storeLoad
 	rankStep *storeLoad
 
-	firstPriority             int
-	secondPriority            int
-	writeLeaderFirstPriority  int
-	writeLeaderSecondPriority int
-	isSelectedDim             func(int) bool
+	// firstPriority and secondPriority indicate priority of hot schedule
+	// they may be byte（0）,key（1），query（2），and always less than dimLen
```
Comment (rleungx) [inline server/schedulers/hot_region.go:440]:
```suggestion
	// they may be byte(0), key(1), query(2), and always less than dimLen
```

File: server/schedulers/hot_region.go [hunk 5]
```diff
@@ -501,18 +502,37 @@ func (bs *balanceSolver) init() {
...
 	}
 
 	bs.isSelectedDim = func(dim int) bool {
 		return dim == bs.firstPriority || dim == bs.secondPriority
 	}
 }
 
+// adjustConfig will adjust config for cluster with low version tikv
+// because tikv below 5.2.0 does not report query information, we will use byte and key as the scheduling dimensions
+func (bs *balanceSolver) adjustConfig(origins, defaults []string) (first, second int) {
```
Comment (rleungx) [inline server/schedulers/hot_region.go:524]: Should we check the length of origins?

File: server/schedulers/hot_region.go [hunk 6]
```diff
@@ -501,18 +502,37 @@ func (bs *balanceSolver) init() {
...
+	case write:
+		switch bs.opTy {
+		case transferLeader:
+			bs.firstPriority, bs.secondPriority = bs.adjustConfig(bs.sche.conf.GetWriteLeaderPriorites(), []string{KeyPriority, BytePriority})
+		case movePeer:
+			bs.firstPriority, bs.secondPriority = bs.adjustConfig(bs.sche.conf.GetWritePeerPriorites(), []string{BytePriority, KeyPriority})
+		}
 	}
 
 	bs.isSelectedDim = func(dim int) bool {
```
Comment (rleungx) [inline server/schedulers/hot_region.go:517]: BTW, seem this function can be moved out from balanceSolver.

File: server/schedulers/hot_region.go (hunk 5 above)
Comment (lhy1024) [reply server/schedulers/hot_region.go:523]: I will do it in another pr

**Review by rleungx — APPROVED** [review]
The rest LGTM except for https://github.com/tikv/pd/pull/3947#discussion_r682421175

**Pull request merged**